crates/ruma-state-res/tests/it/resolve.rs

```diff
@@ -318,6 +339,127 @@ where
     resolve(rules, &state_sets, auth_chain_sets, |x| pdus_by_id.get(x).cloned()).map_err(Into::into)
 }
 
+/// Perform state resolution on a batch of PDUs iteratively, one-by-one.
+///
+/// This function walks the `prev_events` of each PDU forward, resolving each pdu against the
+/// state(s) of it's `prev_events`, to emulate what would happen in a regular room a server is
+/// participating in.
+///
+/// # Arguments
+///
+/// * `auth_rules`: The authorization rules of the room version.
+/// * `pdus`: An iterator of [`Pdu`]s to resolve, with the following assumptions:
+///   * `prev_events` of each PDU points to another provided state event.
+///
+/// # Returns
+///
+/// The state resolved by resolving all the leaves (PDUs which don't have any other PDUs pointing
+/// to it via `prev_events`).
+fn resolve_iteratively<'a, I, II>(
+    auth_rules: &AuthorizationRules,
+    pdus: II,
+) -> Result<StateMap<OwnedEventId>, Box<dyn Error>>
+where
+    I: Iterator<Item = &'a Pdu>,
+    II: IntoIterator<IntoIter = I> + Clone,
+{
+    let mut forward_prev_events_graph: HashMap<&_, Vec<_>> = HashMap::new();
+    let mut stack = Vec::new();
+
+    for pdu in pdus.clone() {
+        let mut has_prev_events = false;
+        for prev_event in pdu.prev_events() {
+            forward_prev_events_graph.entry(prev_event).or_default().push(pdu.event_id());
+            has_prev_events = true;
+        }
+        if pdu.event_type() == &TimelineEventType::RoomCreate && !has_prev_events {
+            stack.push(pdu.event_id().to_owned());
+        }
+    }
+
+    let pdus_by_id: HashMap<OwnedEventId, Pdu> = HashMap::from_iter(
+        pdus.clone().into_iter().map(|pdu| (pdu.event_id().to_owned(), pdu.to_owned())),
+    );
+
+    let mut state_at_events: HashMap<OwnedEventId, StateMap<OwnedEventId>> = HashMap::new();
+    let mut leaves = Vec::new();
+
+    'outer: while let Some(event_id) = stack.pop() {
+        let mut states_before_event = Vec::new();
+        let mut auth_chain_sets = Vec::new();
+
+        let current_pdu = pdus_by_id.get(&event_id).expect("every pdu should be available");
+
+        for prev_event in current_pdu.prev_events() {
+            let Some(state_at_event) = state_at_events.get(prev_event) else {
+                // State for a prev event is not known, we will come back to this event on a later
```

Comment:
If one of the events in `E.prev_events` is missing from `pdus`, then currently we will just silently skip resolving `E` and any events that depend on `E`. Perhaps we should throw an error in this situation instead, to make future debugging easier.